src/dht_proxy_client.cpp

```diff
@@ -221,136 +244,135 @@ time_point
-
-                if (code == 200) {
-                    try {
-                        while (restbed::Http::is_open(req) and not *finished and not state->stop) {
-                            restbed::Http::fetch("\n", reply);
-                            if (*finished or state->stop)
-                                break;
-                            std::string body;
-                            reply->get_body(body);
-                            reply->set_body(""); // Reset the body for the next fetch
-
-                            std::string err;
-                            Json::Value json;
-                            Json::CharReaderBuilder rbuilder;
-                            auto* char_data = reinterpret_cast<const char*>(&body[0]);
-                            auto reader = std::unique_ptr<Json::CharReader>(rbuilder.newCharReader());
-                            if (reader->parse(char_data, char_data + body.size(), &json, &err)) {
-                                auto value = std::make_shared<Value>(json);
-                                if ((not filter or filter(*value)) and cb) {
-                                    {
-                                        std::lock_guard<std::mutex> lock(lockCallbacks);
-                                        callbacks_.emplace_back([cb, value, state]() {
-                                            if (not state->stop and not cb({value}))
-                                                state->stop = true;
-                                        });
-                                    }
-                                    loopSignal_();
+    if (logger_)
+        logger_->d("[proxy:client] [get] [search %s]", key.to_c_str());
+
+    auto request = std::make_shared<http::Request>(httpContext_, resolver_, logger_);
+    auto reqid = request->id();
+    try {
+        restinio::http_request_header_t header;
+        header.request_target("/" + key.toString());
+        header.method(restinio::http_method_get());
+        request->set_header(header);
+        setHeaderFields(request);
+
+        auto state = std::make_shared<OperationState>();
+        Value::Filter filter = w.empty() ? f : f.chain(w.getFilter());
+
+        request->add_on_status_callback([this, key, state](unsigned int status_code){
+            if (status_code != 200){
+                if (logger_)
+                    logger_->e("[proxy:client] [get %s] status error: %i", key.to_c_str(), status_code);
+                state->ok.store(false);
+            }
+        });
+        request->add_on_body_callback([this, key, state, filter, cb](const char* at, size_t length){
+            try {
+                Json::CharReaderBuilder rbuilder;
+                auto body = std::string(at, length);
+                std::vector<dht::Sp<dht::Value>> values;
+                // one value per body line
+                std::string data_line;
+                std::stringstream body_stream(body);
```

Comment:
This has to rely on async_read_until somehow (like restbed::Http::fetch used to do).
There is no guarantee values will arrive in clean blocks. This will likely be an issue for large values (larger than MTU).

Reply:
Unless we're talking about a case where the data size is larger than ```size_t``` which is containing the ```Content-Length``` value, I think it works if we consider this [http example](https://github.com/binarytrails/various/blob/master/cpp/opendht/http.cpp) executed with ```./http.out ns.jami.net 80 /name/binarytrails``` providing ```Content-Length: 1857``` which is greater than it can process in one socket read. Therefore, the http client continuously reads until this length is reached in [handle_response_body](https://github.com/binarytrails/opendht/blob/proxy_restinio/src/http.cpp#L662) which has a ```chunk``` as argument containing the previous read fragment!